### DISCUSSION TITLE
Testing idea that I could work but I'd like to get your thoughts first

### DISCUSSION THREAD

**Discussion by driechers** [mbridak/not1mm#162] [Ideas]
I thought it would be helpful for me to contribute some testing to not1mm. It would be really cool to see some sort of ci pipeline type thing built in so pull requests need to pass existing unit tests. I think it would also be cool to add some QT gui tests. My thought is just start with a few simple usage tests. ie start a contest generate a log, run a few qsos. I know QT has some ability to simulate gui events and I would expect this to be tied into python bindings. Another thought i've had is to get N1MM running in wine and having not1mm and n1mm networked together in some sort of CI test. I understand from your interview in QSO today that keeping these compatible is no easy task but I wonder if a few automated tests in a ci pipleine could make that more realistic? What do you think? I'm eager to start playing with this code. Anything in this approach you would change or prioritize?

**Comment by mbridak** [comment]
I've been thinking about this for awhile. As the code base gets bigger and bigger it's starting to get difficult to keep it all in my head at once. In the back of my head as I make more changes and additions, I'm thinking I hope this doesn't break something existing. I must admit, alot of it could have been avoided if I had taken care in the beginning to carefully craft classes with better defined roles. Instead I just kept bolting things on and as long as they worked I went with it. I've picked up two books on the subject, "Python Testing with pytest" by Brian Okken and "Crafting Test Driven Software" by Alessandro Molina. Of course haven't gotten around to reading them. The two most neglected things in software are tests and documentation. I'd be very open to anything you'd want to tackle in this area.

**Reply by driechers** [reply]
Sounds great! I’m thinking I’ll start by running existing tests and fix if needed then make a ci hook. I’ve played around with this a bit in the past with GitHub and gitlab. Shouldn’t be too hard to get those neurons fired back up. I don’t recall how access control worked Jo that so I’ll likely have to get back to you on this thread since you would want to be admin on any resources. I’m excited this will be fun!

**Reply by driechers** [reply]
I got some stuff working last night. Hopefully making a pr by this weekend. I also realized i probably heard you on linux in the ham shack and not qso today haha got a little mixed up.

**Comment by mbridak** [comment]
That's okay. My brain auto-corrected and I didn't notice.

**Comment by driechers** [comment]
Testing going good. I have a good start with pytestqt. I'm filling out test data for every contest now. I am finding bugs and fixing along the way!

**Comment by mbridak** [comment]
Excellent, I'll be interested to see how it works.

**Reply by driechers** [reply]
Trying to stop myself from being a perfectionist. Already want to refactor all my code.

**Comment by driechers** [comment]
Opened up #165 I was trying to fix all the bugs I was finding but was struggling with a couple and figured I’d share what I have. Working during kids nap times so time is fairly limited.

**Reply by mbridak** [reply]
Ah nap times. I remember those. Mine are in Jr. High and High school now. Even though I got to retire early I still have a fairly large chunk of the day shuttling kids to school, friends, soccer practices etc. But totally worth it.

Enjoy the quiet times.

**Comment by mbridak** [comment]
Looks good. I merged it. Thanks for the PR!

**Comment by mbridak** [comment]
If I'm not mistaken... Most of the errors you might be seeing stem from there is no Station information set in the beginning of the test.
It looks like the test DB is created, then the Settings dialog that comes up is just closed without any info being put into it.
None of the text in the fields you see when the station settings dialog is opened is saved. It's just placeholder text to give the user an idea as to what the field is expecting.

**Reply by driechers** [reply]
Yeah I was seeing in the code some get calls were a default option is provided if it is not set. But I think settings is making it an empty string instead of None so default is not being taken. The itu zone errors are for sure from that. The cty errors are a bit trickier for me to wrap my head around. I think what is happening is the search is turning up nothing not sure. My thought is nothing should make it crash so perhaps interface should have required fields or default fields filled in so it can’t get in a bad state. Not sure I understand the bugs all yet

**Comment by mbridak** [comment]
I've added the info to the test and it looks like it's passing now.

```bash
=============================================================== test session starts ===============================================================
platform linux -- Python 3.12.6, pytest-8.3.3, pluggy-1.5.0
PyQt6 6.7.1 -- Qt runtime 6.7.3 -- Qt compiled 6.7.1
rootdir: /home/mbridak/Nextcloud/dev/not1mm
configfile: pyproject.toml
plugins: qt-4.4.0
collected 1 item                                                                                                                                  

test/contests.py .                                                                                                                          [100%]

=============================================================== 1 passed in 22.10s ================================================================
(venv) mbridak@fredo:~/Nextcloud/dev/not1mm$ 
```

**Reply by driechers** [reply]
Awesome! I'll play around with getting it working headless and put together a dockerfile. https://github.com/features/actions this seems to be the way to get it running in a pipeline.

**Reply by driechers** [reply]
Getting this working headlessly is proving harder than I expected. Still looking into it. Just slow cause of limited time

**Comment by mbridak** [comment]
No worries. This is a zero stress/obligation environment.
